presto-hive/src/main/java/com/facebook/presto/hive/orc/HdfsOrcDataSource.java

```diff
@@ -73,6 +73,9 @@ protected void readInternal(long position, byte[] buffer, int bufferOffset, int
             if (e.getClass().getSimpleName().equals("BlockMissingException")) {
                 throw new PrestoException(HIVE_MISSING_DATA, message, e);
             }
+            if (e instanceof IOException) {
```

Comment:
Maybe

```java
@Override
protected void readInternal(long position, byte[] buffer, int bufferOffset, int bufferLength)
{
    try {
        long readStart = System.nanoTime();
        inputStream.readFully(position, buffer, bufferOffset, bufferLength);
        stats.readDataBytesPerSecond(bufferLength, System.nanoTime() - readStart);
    }
    catch (PrestoException e) {
        // just in case there is a Presto wrapper or hook
        throw e;
    }
    catch (BlockMissingException e) {
        throw new PrestoException(HIVE_MISSING_DATA, formatReadErrorMessage(position), e);
    }
    catch (IOException e) {
        throw new PrestoException(HIVE_FILESYSTEM_ERROR, formatReadErrorMessage(position), e);
    }
    catch (Exception e) {
        throw new PrestoException(HIVE_UNKNOWN_ERROR, formatReadErrorMessage(position), e);
    }
}

private String formatReadErrorMessage(long position)
{
    return format("Error reading from %s at position %s", this, position);
}

Reply:
BlockMissingException is in different packages depending on the Hadoop version.

Reply:
So, unfortunately, this needs to stay as-is because `BlockMissingException` extends `IOException` and thus the check for a general `IOException` needs to come afterwards.

Though I've been thinking we should just remove `BlockMissingException` handling since it can mean two different things (transient HDFS error or permanent data loss) and is also HDFS specific.